AsciiPixel/src/ascii.jl

```diff
@@ -0,0 +1,244 @@
+abstract type ImageEncoder end
+struct BigBlocks <: ImageEncoder
+    size::NTuple{2, Int}
+end
+struct SmallBlocks <: ImageEncoder
+    size::NTuple{2, Int}
+end
+
+const RESET = Crayon(reset = true)
+const alpha_chars = ('⋅', '░', '▒', '▓', '█')
+
+function _charof(alpha)
+    idx = round(Int, alpha * (length(alpha_chars)-1))
+    alpha_chars[clamp(idx + 1, 1, length(alpha_chars))]
+end
+
+"""
+    downscale_XXX(img::AbstractMatrix{<:Colorant}, maxheight::Int, maxwidth::Int)
+
+Larger images are downscaled automatically using `restrict`.
+
+- `maxheight` and `maxwidth` specify the maximum numbers of
+  string characters that should be used for the resulting image.
+
+Returns
+    1. Downscaled image
+    2. Selected encoder with `size` containing:
+        1. number of lines in the vector.
+        2. number of visible characters per line (the remaining are colorcodes).
+"""
+function downscale_small(img::AbstractMatrix{<:Colorant}, maxheight::Int, maxwidth::Int)
+    maxheight = max(maxheight, 5)
+    maxwidth  = max(maxwidth,  5)
+    h, w = map(length, axes(img))
+    while ceil(h/2) > maxheight || w > maxwidth
+        img = restrict(img)
+        h, w = map(length, axes(img))
+    end
+    img, SmallBlocks((length(1:2:h), w))
+end
```

Comment:
There can be two possible `restrict` enhancements in the future:

- an n-fold restrict so that we can do it in one call `restrict(img; n=downscale_ratio)` https://github.com/JuliaImages/ImageBase.jl/issues/30
- a `restrict_to(img, sz)` method that recursively call `restrict` until the image size is strictly <= `sz`.

If we take this into consideration, maybe it's better to

```julia
function _restrict_to(img::AbstractMatrix, max_sz)
    while any(size(img) .> max_sz)
        img = restrict(img)
    end
    return img
end

function ascii_display(...)
    ...
    # the arithmetics here might be incorrect
    if img_h <= io_h - 4 && 2img_w <= io_w
        ...
    else
        img = _restrict_to(img, max_sz .* (2, 1))
        enc = SmallBlocks(max_sz .÷ (2, 1))
    end
    ...
    return io
end
```

so that in the future we can directly replace `_restrict_to` with the better version provided by ImageBase. Also we do not need to document the `downscale_xxx` functions, but just leave a few comments on how the numbers are calculated.

How do you think?

Reply:
That would require to change the tests, dropping testing the downscaling size, but OK.

Reply:
I'm gonna change my mind about this. This requires to much changes and the function `_restrict_to` does not take into account the current multiple specific arithmetics, thus is not equivalent to the current code e.g.:

https://github.com/JuliaImages/ImageInTerminal.jl/blob/4cd5bebf718196cb49a66b600b27d4023696ead4/src/encodeimg.jl#L81

versus

https://github.com/JuliaImages/ImageInTerminal.jl/blob/4cd5bebf718196cb49a66b600b27d4023696ead4/src/encodeimg.jl#L48.

I don't see any solution to this, it was already difficult to move logic outside `encodeimg` e.g. for https://github.com/JuliaImages/ImageInTerminal.jl/blob/4cd5bebf718196cb49a66b600b27d4023696ead4/src/encodeimg.jl#L130.

We can change this later.